File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/action/QuitGameAction.java
package org.emulinker.kaillera.controller.v086.action;

import com.google.common.flogger.FluentLogger;
import javax.inject.Inject;
import javax.inject.Singleton;
import org.emulinker.kaillera.controller.messaging.MessageFormatException;
import org.emulinker.kaillera.controller.v086.V086ClientHandler;
import org.emulinker.kaillera.controller.v086.protocol.*;
import org.emulinker.kaillera.lookingforgame.TwitterBroadcaster;
import org.emulinker.kaillera.model.KailleraUser;
import org.emulinker.kaillera.model.event.*;
import org.emulinker.kaillera.model.exception.*;

@Singleton
public class QuitGameAction
    implements V086Action<QuitGame_Request>, V086GameEventHandler<UserQuitGameEvent> {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private static final String DESC = "QuitGameAction";

  private int actionCount = 0;
  private int handledCount = 0;

  private final TwitterBroadcaster lookingForGameReporter;

  @Inject
  QuitGameAction(TwitterBroadcaster lookingForGameReporter) {
    this.lookingForGameReporter = lookingForGameReporter;
  }

  @Override
  public int getActionPerformedCount() {
    return actionCount;
  }

  @Override
  public int getHandledEventCount() {
    return handledCount;
  }

  @Override
  public String toString() {
    return DESC;
  }

  @Override
  public void performAction(QuitGame_Request message, V086ClientHandler clientHandler)
      throws FatalActionException {
    actionCount++;

    try {
      clientHandler.getUser().quitGame();
      lookingForGameReporter.cancelActionsForUser(clientHandler.getUser().getID());
    } catch (DropGameException | QuitGameException | CloseGameException e) {
      logger.atSevere().withCause(e).log("Action failed");
    }

    try {
      Thread.sleep(100);
    } catch (InterruptedException e) {
      logger.atSevere().withCause(e).log("Sleep Interrupted!");
    }
  }

  @Override
  public void handleEvent(UserQuitGameEvent userQuitEvent, V086ClientHandler clientHandler) {
    handledCount++;
    KailleraUser thisUser = clientHandler.getUser();

    try {
      KailleraUser user = userQuitEvent.getUser();

      if (user.getStealth() == false)
        clientHandler.send(
            QuitGame_Notification.create(
                clientHandler.getNextMessageNumber(), user.getName(), user.getID()));

      if (thisUser == user) {
        if (user.getStealth() == true)
          clientHandler.send(
              QuitGame_Notification.create(
                  clientHandler.getNextMessageNumber(), user.getName(), user.getID()));
      }
    } catch (MessageFormatException e) {
      logger.atSevere().withCause(e).log("Failed to contruct QuitGame_Notification message");
    }
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/action/QuitGameAction.kt
package org.emulinker.kaillera.controller.v086.action

import com.google.common.flogger.FluentLogger
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.time.Duration.Companion.milliseconds
import kotlinx.coroutines.delay
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.v086.V086ClientHandler
import org.emulinker.kaillera.controller.v086.protocol.QuitGame_Notification
import org.emulinker.kaillera.controller.v086.protocol.QuitGame_Request
import org.emulinker.kaillera.lookingforgame.TwitterBroadcaster
import org.emulinker.kaillera.model.event.UserQuitGameEvent
import org.emulinker.kaillera.model.exception.CloseGameException
import org.emulinker.kaillera.model.exception.DropGameException
import org.emulinker.kaillera.model.exception.QuitGameException

private val logger = FluentLogger.forEnclosingClass()

@Singleton
class QuitGameAction @Inject constructor(private val lookingForGameReporter: TwitterBroadcaster) :
    V086Action<QuitGame_Request>, V086GameEventHandler<UserQuitGameEvent> {
  override var actionPerformedCount = 0
    private set
  override var handledEventCount = 0
    private set

  override fun toString() = "QuitGameAction"

  @Throws(FatalActionException::class)
  override suspend fun performAction(message: QuitGame_Request, clientHandler: V086ClientHandler) {
    actionPerformedCount++
    try {
      clientHandler.user.quitGame()
      lookingForGameReporter.cancelActionsForUser(clientHandler.user.id)
    } catch (e: DropGameException) {
      logger.atSevere().withCause(e).log("Action failed")
    } catch (e: QuitGameException) {
      logger.atSevere().withCause(e).log("Action failed")
    } catch (e: CloseGameException) {
      logger.atSevere().withCause(e).log("Action failed")
    }
    delay(100.milliseconds)
  }

  override suspend fun handleEvent(event: UserQuitGameEvent, clientHandler: V086ClientHandler) {
    handledEventCount++
    val thisUser = clientHandler.user
    try {
      val user = event.user
      if (!user.inStealthMode) {
        clientHandler.send(
            QuitGame_Notification(clientHandler.nextMessageNumber, user.name!!, user.id))
      }
      if (thisUser === user) {
        if (user.inStealthMode)
            clientHandler.send(
                QuitGame_Notification(clientHandler.nextMessageNumber, user.name!!, user.id))
      }
    } catch (e: MessageFormatException) {
      logger.atSevere().withCause(e).log("Failed to construct QuitGame_Notification message")
    }
  }
}
